src/lib/phdi-building-blocks/phdi_building_blocks/fhir.py

```diff
@@ -125,3 +128,184 @@ def upload_bundle_to_fhir_server(bundle: dict, access_token: str, fhir_url: str)
+    :param fhir_url: FHIR Server base URL
+    :param export_scope: Either `Patient` or `Group/[id]` as specified in the FHIR spec
+    (https://hl7.org/fhir/uv/bulkdata/export/index.html#bulk-data-kick-off-request)
+    :param since: A FHIR instant (https://build.fhir.org/datatypes.html#instant)
+    instructing the export to include only resources created or modified after the
+    specified instant.
+    :param resource_type: A comma-delimited list of resource types to include.  All
+    :param poll_step: the number of seconds to wait between poll requests, waiting
+    for export files to be generated.
+    :param poll_timeout: the maximum number of seconds to wait for export files to
+    be generated.
+    """
+    logging.debug("Initiating export from FHIR server.")
+    export_url = _compose_export_url(
+        fhir_url=fhir_url,
+        export_scope=export_scope,
+        since=since,
+        resource_type=resource_type,
+    )
+    logging.debug(f"Composed export URL: {export_url}")
+    response = requests.get(
+        export_url,
+        headers={
+            "Authorization": f"Bearer {access_token}",
+            "Accept": "application/fhir+json",
+            "Prefer": "respond-async",
+        },
+    )
+    logging.info(f"Export request completed with status {response.status_code}")
+
+    if response.status_code == 202:
+
+        poll_response = export_from_fhir_server_poll(
+            poll_url=response.headers.get("Content-Location"),
+            access_token=access_token,
+            poll_step=poll_step,
+            poll_timeout=poll_timeout,
+        )
+
+        if poll_response.status_code == 200:
+            logging.debug(f"Export content: {poll_response.text}")
+            return poll_response.json()
+        else:
+            logging.exception("Unexpected response code during export download.")
+            raise requests.HTTPError(response=response)
+
+
+def _compose_export_url(
+    fhir_url: str, export_scope: str = "", since: str = "", resource_type: str = ""
+) -> str:
+    """Generate a query string for the export request.  Details in the FHIR spec:
+    https://hl7.org/fhir/uv/bulkdata/export/index.html#query-parameters"""
+    export_url = fhir_url
+    if export_scope == "Patient" or export_scope.startswith("Group/"):
+        export_url += f"/{export_scope}/$export"
+    elif export_scope == "":
+        export_url += "/$export"
+    else:
+        raise ValueError("Invalid scope {scope}.  Expected 'Patient' or 'Group/[ID]'.")
+
+    # Start with ? url argument separator, and change it to & after the first parameter
+    # is appended to the URL
+    separator = "?"
+    if since:
+        export_url += f"{separator}_since={since}"
+        separator = "&"
+
+    if resource_type:
+        export_url += f"{separator}_type={resource_type}"
+        separator = "&"
+
+    return export_url
+
+
+def __export_from_fhir_server_poll_call(
+    export_url: str, access_token: str
```

Comment:
It's a nit because we pass `poll_url` as the arg in the poll function below, but can we change this parameter name to `poll_url`? I was confused when I was first reading through this because export_url is the string we used to actually kick off the export, whereas poll_url is the status checking URL the server sends back in response.